src/utils/src/data-scale-utils.ts

```diff
@@ -55,3 +75,315 @@ export function getLogDomain(data: any[], valueAccessor: dataValueAccessor): [nu
   const [d0, d1] = getLinearDomain(data, valueAccessor);
   return [d0 === 0 ? 1e-5 : d0, d1];
 }
+
+/**
+ * whether field domain is stops
+ */
+export function isDomainStops(domain: any): boolean {
```

Comment:
Could these do type narrowing?
```suggestion
export function isDomainStops(domain: any): domain is DomainWithStops {
```